### PR TITLE
T-16 Create functionality to retrieve data from car flea market

### PR DESCRIPTION
**Pull request by DashaDrahun**
Create functionality to retrieve car data from onliner

### CODE REVIEW

File: PortfolioKeeper/BusinessLogic/Services/CarInfoService.cs [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+using PortfolioKeeper.BusinessLogic.Parsers.Cars;
+using PortfolioKeeper.BusinessLogic.Providers;
+using PortfolioKeeper.Contracts.Enums;
+using PortfolioKeeper.CustomExceptions;
+using PortfolioKeeper.Data.Entities;
+using PortfolioKeeper.Models.Dto;
+
+namespace PortfolioKeeper.BusinessLogic.Services
+{
+    public class CarInfoService : ICarInfoService
```
Comment (kokarekom) [inline PortfolioKeeper/BusinessLogic/Services/CarInfoService.cs]: Too much logic inside for a single public method - put it to appropriate helpers to keep it clear.

File: PortfolioKeeper/BusinessLogic/Services/PortfolioValuationService.cs [hunk 2]
```diff
@@ -59,5 +59,10 @@ public IEnumerable<PortfolioValuationDto> Get(string userId, int id, TimeRange r
 
             return mapper.Map<IEnumerable<PortfolioValuation>, IEnumerable<PortfolioValuationDto>>(valuations);
         }
+
+        public void Create(Portfolio asset, decimal value)
+        {
+            throw new NotImplementedException();
```
Comment (kokarekom) [inline PortfolioKeeper/BusinessLogic/Services/PortfolioValuationService.cs]: ?

File: PortfolioKeeper/Contracts/Enums/PropertyType.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿namespace PortfolioKeeper.Contracts.Enums
+{
+    public enum PropertyType
```
Comment (kokarekom) [inline PortfolioKeeper/Contracts/Enums/PropertyType.cs]: Why mix of different entities? Please set a more specific name to the enum.

File: PortfolioKeeper/BusinessLogic/Services/AssetService.cs [hunk 4]
```diff
@@ -6,18 +6,20 @@
...
 using PortfolioKeeper.Models.Dto;
 using PortfolioKeeper.Models.ViewModels;
+using WebGrease.Css.Extensions;
 
 namespace PortfolioKeeper.BusinessLogic.Services
 {
     public class AssetService : IAssetService
     {
-        private static readonly Logger logger = LogManager.GetCurrentClassLogger();
+        private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
```
Comment (kokarekom) [inline PortfolioKeeper/BusinessLogic/Services/AssetService.cs]: Variable name to a lower camel case.

File: PortfolioKeeper/BusinessLogic/Services/PortfolioValuationService.cs (hunk 2 above)
Comment (DashaDrahun) [reply PortfolioKeeper/BusinessLogic/Services/PortfolioValuationService.cs]: This method will be used when we implement portfolio valuation functionality (task T-22)

File: PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/ICarModelService.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+﻿using System.Collections.Generic;
+using PortfolioKeeper.BusinessLogic.Parsers.Cars;
+using PortfolioKeeper.Data.Entities;
+
+namespace PortfolioKeeper.BusinessLogic.Services.RetrievingCarPrice
+{
+    public interface ICarModelService
+    {
+        double GetBestPrice(IEnumerable<CarJsonModel> brandsWithModels, Asset asset, string brandId);
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/ICarModelService.cs]: brandsWithModels-> models

File: PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarModelService.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        {
+            this.xmlParser = xmlParser
+                             ?? throw new ArgumentNullException(nameof(xmlParser));
+            this.modelHelper = modelHelper
+                             ?? throw new ArgumentNullException(nameof(modelHelper));
+        }
+
+        public double GetBestPrice (IEnumerable<CarJsonModel> brandsWithModels, Asset asset, string brandId)
+        {
+            if (asset == null)
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarModelService.cs]: change args order - asset, brandId, brandWithModels

File: PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarModelService.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                throw new ArgumentNullException(nameof(asset));
+            if (brandsWithModels == null)
+                throw new ArgumentNullException(nameof(brandsWithModels));
+            if (brandId.IsNullOrWhiteSpace())
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarModelService.cs]: brandId.IsNullOrWhiteSpace() -> string.IsNullOrWhiteSpace()

File: PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarModelService.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+                throw new ArgumentNullException(nameof(brandId));
+
+            var brands = brandsWithModels.ToList();
+            var modelId = modelHelper.FindCarModelId(brands, asset);
+
+            if (modelId.IsNullOrWhiteSpace())
+            {
+                logger.Warn($"Model of asset (assetId: {asset.Id} didn't find in json");
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarModelService.cs]: assetId: {asset.Id} -> assetId = {asset.Id}.  Change log - Model could not be found. AssetId = {asset.Id}.

File: PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarModelService.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        public CarModelService(IModelHelper modelHelper,
+            IXmlParser xmlParser)
+        {
+            this.xmlParser = xmlParser
+                             ?? throw new ArgumentNullException(nameof(xmlParser));
+            this.modelHelper = modelHelper
+                             ?? throw new ArgumentNullException(nameof(modelHelper));
+        }
+
+        public double GetBestPrice (IEnumerable<CarJsonModel> brandsWithModels, Asset asset, string brandId)
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarModelService.cs]: better make double nullable and return null not  0

File: PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarModelService.cs [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+            {
+                logger.Warn($"Model of asset (assetId: {asset.Id} didn't find in json");
+                return 0;
+            }
+
+            var models = modelHelper.GetJson(brandId, modelId);
+
+            if (models == null)
+            {
+                logger.Warn($"Request didn't find models of brand (brandId: {brandId})");
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarModelService.cs]: Brand models could not be found. BrandId = {brandId}.

File: PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarModelService.cs [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+                return 0;
+            }
+
+            var modelsWithYear = models.Where(x =>
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarModelService.cs]: modelsWithYear -> filteredModels

File: PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarModelService.cs [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+                x.Year.Year == xmlParser.ParsePropertyFromXml(asset.PropertyValues.FirstOrDefault(y => y.TypeId == (int)PropertyTypeName.Year)?.Value
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarModelService.cs]: too complicated. better get car year in one line, then search through models

File: PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarInfoService.cs [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+                foreach (var asset in filteredAssets)
+                {
+                    var bestPrice = modelService.GetBestPrice(brandsWithModels, asset, brand.Id);
+                    assetService.UpdateCarPrice(bestPrice, asset);
+                }
+            }
+        }
+    }
+}
+
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarInfoService.cs]: remove empty lines

File: PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarInfoService.cs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+            foreach (var brand in brands)
+            {
+                var filteredAssets = brandHelper.GetAssetsFilteredByBrand(brand.Name, assets);
+                var brandsWithModels = brandHelper.GetBrandsWithCarModels(brand.Id).ToList();
+
+                foreach (var asset in filteredAssets)
+                {
+                    var bestPrice = modelService.GetBestPrice(brandsWithModels, asset, brand.Id);
+                    assetService.UpdateCarPrice(bestPrice, asset);
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarInfoService.cs]: if price =0 or null after code changes do we really need to update price? there should be check

File: PortfolioKeeper/BusinessLogic/Services/IAssetService.cs [hunk 15]
```diff
@@ -31,5 +31,11 @@ void AddConcreteAsset<T>(T asset, byte childAssetTypeId, string userId)
         void Update(AssetDetailsViewModel model);
 
         Asset GetAsset(int assetId);
+
+        IList<Asset> GetAllCars();
+
+        void Update(Asset model);
+
+        void UpdateCarPrice(double bestPrice, Asset asset);
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/IAssetService.cs]: be constistent - if you have Create(Asset asset, decimal value) then change  UpdateCarPrice(double bestPrice, Asset asset); to  UpdateCarPrice(Asset asset, double bestPrice);

File: PortfolioKeeper/BusinessLogic/Services/AssetService.cs [hunk 16]
```diff
@@ -175,6 +182,13 @@ public Asset GetAsset(int assetId)
             throw new ResourceNotFoundException($"Asset not found (assetId: {assetId})");
         }
 
+        public IList<Asset> GetAllCars()
+        {
+            var cars = assetRepository.Get(e => e.AssetTypeId == (int)AssetTypeValue.Car, null, "PropertyValues");
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/AssetService.cs]: "PropertyValues" - magic string? add to constants, please, if there is no

File: PortfolioKeeper/BusinessLogic/Services/AssetService.cs [hunk 17]
```diff
@@ -288,6 +302,16 @@ public void Update(AssetDetailsViewModel model)
             unitOfWork.Commit();
         }
 
+        public void Update(Asset model)
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/AssetService.cs]: model -> asset

File: PortfolioKeeper/BusinessLogic/Services/AssetService.cs [hunk 18]
```diff
@@ -317,5 +341,31 @@ public IEnumerable<AssetPropertyValueDto> GetAssetPropertyValues(int assetId, st
...
+            if (asset == null)
+                throw new ArgumentNullException(nameof(asset));
+            if (Math.Abs(bestPrice) < 0.01)
+            {
+                logger.Warn($"Price for assetId: {asset.Id} not found");
+                return;
+            }
+
+            var property = asset.PropertyValues.FirstOrDefault(x => x.TypeId == (int)PropertyTypeName.VehiclePrice);
+            if (property != null)
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/AssetService.cs]: add line before if

File: PortfolioKeeper/BusinessLogic/Services/AssetService.cs [hunk 19]
```diff
@@ -317,5 +341,31 @@ public IEnumerable<AssetPropertyValueDto> GetAssetPropertyValues(int assetId, st
...
+            {
+                if (Convert.ToDecimal(xmlParser.ParsePropertyFromXml(property.Value)) == (decimal)bestPrice)
+                    return;
+                property.Value = xmlParser.ParsePropertyToXml((int)PropertyValueType.Decimal,
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/AssetService.cs]: add line after return

File: PortfolioKeeper/BusinessLogic/Providers/CarInfoProvider.cs [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            string url = config.OnlinerCarsLinks.ModelsLink + id;
+
+            return GetResponse(url).Result;
+        }
+
+        public string GetModelInfo(string brand, string model)
+        {
+            string url = config.OnlinerCarsLinks.AnnouncementLink;
+
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Providers/CarInfoProvider.cs]: remove empty line

File: PortfolioKeeper/BusinessLogic/Providers/CarInfoProvider.cs [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+        public CarInfoProvider(IRequestProvider requestProvider, IOnlinerConfiguration config)
+        {
+            this.requestProvider = requestProvider
+                                   ?? throw new ArgumentNullException(nameof(requestProvider));
+            this.config = config
+                                   ?? throw new ArgumentNullException(nameof(config));
+        }
+
+        private async Task<string> GetResponse(string url)
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Providers/CarInfoProvider.cs]: this should be moved from here. use or extend RequestProvider

File: PortfolioKeeper/BusinessLogic/Providers/CarInfoProvider.cs [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            if (String.IsNullOrWhiteSpace(url))
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Providers/CarInfoProvider.cs]: String.IsNullOrWhiteSpace -> string.IsNullOrWhiteSpace

File: PortfolioKeeper/BusinessLogic/Providers/CarInfoProvider.cs [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+using PortfolioKeeper.CustomExceptions;
+
+namespace PortfolioKeeper.BusinessLogic.Providers
+{
+    public class CarInfoProvider : ICarInfoProvider
+    {
+        private readonly IRequestProvider requestProvider;
+        private readonly IOnlinerConfiguration config;
+
+        public CarInfoProvider(IRequestProvider requestProvider, IOnlinerConfiguration config)
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Providers/CarInfoProvider.cs]: IOnlinerConfiguration config to new line

File: PortfolioKeeper/BusinessLogic/Parsers/Cars/JsonModelsForCarModels/YearModel.cs [hunk 24]
```diff
@@ -0,0 +1,10 @@
+﻿using Newtonsoft.Json;
+
+namespace PortfolioKeeper.BusinessLogic.Parsers.Cars.JsonModelsForCarModels
+{
+    public class YearModel
+    {
+        [JsonProperty("year")]
+        public string Year { get; set; }
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Parsers/Cars/JsonModelsForCarModels/YearModel.cs]: rename to Value

File: PortfolioKeeper/BusinessLogic/Parsers/CarInfoParser.cs [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+        public CarInfoParser(ICarSerializer serializer)
+        {
+            this.serializer = serializer ??
+                              throw new ArgumentNullException(nameof(serializer));
+        }
+
+        public IEnumerable<M> ParseAllCars<T, M>(string response)
+            where T : Wrapper<M>
+        {
+            if (response.IsNullOrWhiteSpace())
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Parsers/CarInfoParser.cs]: response.IsNullOrWhiteSpace() -> string.IsNullOrWhiteSpace()

File: PortfolioKeeper/BusinessLogic/Helpers/RetrievingCarPrice/ModelHelper.cs [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+        {
+            this.provider = provider
+                            ?? throw new ArgumentNullException(nameof(provider));
+            this.carParser = carParser
+                             ?? throw new ArgumentNullException(nameof(carParser));
+            this.xmlParser = xmlParser
+                             ?? throw new ArgumentNullException(nameof(xmlParser));
+        }
+
+        public string FindCarModelId(IEnumerable<CarJsonModel> brandsWithModels, Asset asset)
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Helpers/RetrievingCarPrice/ModelHelper.cs]: better pass asset.PropertyValues here. no need to pass full object

File: PortfolioKeeper/BusinessLogic/Helpers/RetrievingCarPrice/ModelHelper.cs [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+            return brandsWithModels.Where(x =>
+                    x.Name == xmlParser.ParsePropertyFromXml(
+                        asset.PropertyValues.FirstOrDefault(y => y.TypeId == (int)PropertyTypeName.Model)?.Value
+                        ?? throw new InvalidOperationException()))
+                .Select(m => m.Id).FirstOrDefault();
+        }
+
+        public IEnumerable<ModelJsonModel> GetJson(string brandId, string modelId)
+        {
+            if (brandId.IsNullOrWhiteSpace())
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Helpers/RetrievingCarPrice/ModelHelper.cs]: brandId.IsNullOrWhiteSpace() -> string.IsNullOrWhiteSpace

File: PortfolioKeeper/App_Start/Startup.cs [hunk 28]
```diff
@@ -30,9 +31,15 @@ public void Configuration(IAppBuilder app)
             app.UseHangfireDashboard();
             app.UseHangfireServer();
 
+            container.Resolve<ICarInfoService>().ProcessCurrentCarValue();
```
Comment (stas-litvinenka) [inline PortfolioKeeper/App_Start/Startup.cs]: remove, as Andrew suggested. it useless if you look this file through

File: PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarInfoService.cs (hunk 14 above)
Comment (DashaDrahun) [reply PortfolioKeeper/BusinessLogic/Services/RetrievingCarPrice/CarInfoService.cs]: I check bestPrice for 0 or null in UpdateCarPrice
            if (asset == null)
                throw new ArgumentNullException(nameof(asset));
            if (bestPrice == null || Math.Abs((double)bestPrice) < 0.01)
            {
                logger.Warn($"Price for assetId: {asset.Id} not found");
                return;
            }

File: PortfolioKeeper/BusinessLogic/Providers/CarInfoProvider.cs (hunk 21 above)
Comment (DashaDrahun) [reply PortfolioKeeper/BusinessLogic/Providers/CarInfoProvider.cs]: But this method uses RequestProvider (it calls GetAsync from RequestProvider  service)

**Pull request merged**